src/Tools/ExternalAccess/VisualDiagnostics/Internal/VisualDiagnosticsServiceFactory.cs

```diff
@@ -0,0 +1,103 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.BrokeredServices;
+using Microsoft.CodeAnalysis.ExternalAccess.VisualDiagnostics.Contracts;
+using Microsoft.CodeAnalysis.ExternalAccess.VisualDiagnostics.Internal;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.ServiceHub.Framework;
+using Roslyn.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VisualDiagnostics;
+
+/// <summary>
+/// LSP Service responsible for loading IVisualDiagnosticsLanguageService workspace service and delegate the broker service to the workspace service,
+/// and handling MAUI XAML/C#/CSS/Razor Hot Reload support
+/// </summary>
+[ExportCSharpVisualBasicLspServiceFactory(typeof(OnInitializedService)), Shared]
+internal sealed class VisualDiagnosticsServiceFactory : ILspServiceFactory
+{
+    private readonly LspWorkspaceRegistrationService _lspWorkspaceRegistrationService;
+    private readonly VisualDiagnosticsServiceBroker _brokeredDebuggerServices;
+
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    [ImportingConstructor]
+    public VisualDiagnosticsServiceFactory(
+        LspWorkspaceRegistrationService lspWorkspaceRegistrationService,
+        VisualDiagnosticsServiceBroker brokeredDebuggerServices)
+    {
+        _lspWorkspaceRegistrationService = lspWorkspaceRegistrationService;
+        _brokeredDebuggerServices = brokeredDebuggerServices;
+    }
+
+    public ILspService CreateILspService(LspServices lspServices, WellKnownLspServerKinds serverKind)
+    {
+        return new OnInitializedService(_lspWorkspaceRegistrationService, _brokeredDebuggerServices);
+    }
+
+    private class OnInitializedService : ILspService, IOnInitialized, IOnServiceBrokerInitialized, IDisposable
+    {
+        private readonly LspWorkspaceRegistrationService _lspWorkspaceRegistrationService;
+        private readonly VisualDiagnosticsServiceBroker _brokeredDebuggerServices;
+        private IVisualDiagnosticsLanguageService? _visualDiagnosticsLanguageService;
+        private CancellationToken _cancellationToken;
+        private static readonly TaskCompletionSource<bool> _taskCompletionSource = new TaskCompletionSource<bool>();
+
+        public OnInitializedService(LspWorkspaceRegistrationService lspWorkspaceRegistrationService, VisualDiagnosticsServiceBroker brokeredDebuggerServices)
+        {
+            _lspWorkspaceRegistrationService = lspWorkspaceRegistrationService;
+            _brokeredDebuggerServices = brokeredDebuggerServices;
+
+            _brokeredDebuggerServices.NotifyServiceBrokerInitialized = this;
```

Comment:
I think there might be a slightly cleaner way to do this
1.  Add another export to `VisualDiagnosticsServiceFactory` to export it as a `IOnServiceBrokerInitialized` service.  Since the export is marked as shared we'll only have 1 instance even if we export it twice.
2. Implement `IOnServiceBrokerInitialized` inside `VisualDiagnosticsServiceFactory` (and delete `VisualDiagnosticsServiceBroker`).  The implementation would call the instance of `OnInitializedService` that gets created in `CreateILspService` (assign it to a field there).

Reply:
Yeah good Idea, I worked! I thought that IOnServiceBrokerInitialized  export from VisualDiagnosticsServiceFactory would be imcompatible with ServiceBrokerFactory since one uses System.ComponentModel.Composition and the other uses System.Composition